hw4/hw4.py

```diff
@@ -56,75 +43,59 @@ def lr(df):
-    X, y, test_size=0.3, random_state=42
-)
-forest = RandomForestClassifier(random_state=0)
-forest.fit(X_train, y_train)
-import time
+    # diabetes = datasets.load_diabetes(as_frame=True)
+    # df = diabetes['frame']
+    X = df.iloc[:, :-1]
+    y = df.iloc[:, -1]
 
-import numpy as np
+    X_train, X_test, y_train, y_test = train_test_split(
+        X, y, test_size=0.3, random_state=42
+    )
+    forest = RandomForestClassifier(random_state=0)
+    forest.fit(X_train, y_train)
+
+    start_time = time.time()
+    importances = forest.feature_importances_
+    std = np.std([tree.feature_importances_ for tree in forest.estimators_], axis=0)
+    elapsed_time = time.time() - start_time
+    print(f"Elapsed time to compute the importances: {elapsed_time:.3f} seconds")
+    forest_importances_impurity = pd.Series(importances, index=df.iloc[:, :-1].columns)
+
+    fig, ax = plt.subplots()
+    forest_importances_impurity.plot.bar(yerr=std, ax=ax)
+    ax.set_title("Feature importances using MDI")
+    ax.set_ylabel("Mean decrease in impurity")
+    fig.tight_layout()
+
+    start_time = time.time()
+    result = permutation_importance(
+        forest, X_test, y_test, n_repeats=10, random_state=42, n_jobs=2
+    )
+    elapsed_time = time.time() - start_time
+    print(f"Elapsed time to compute the importances: {elapsed_time:.3f} seconds")
 
-start_time = time.time()
-importances = forest.feature_importances_
-std = np.std([tree.feature_importances_ for tree in forest.estimators_], axis=0)
-elapsed_time = time.time() - start_time
-print(f"Elapsed time to compute the importances: {elapsed_time:.3f} seconds")
-forest_importances_impurity = pd.Series(importances, index=df.columns[0:5])
-
-fig, ax = plt.subplots()
-forest_importances_impurity.plot.bar(yerr=std, ax=ax)
-ax.set_title("Feature importances using MDI")
-ax.set_ylabel("Mean decrease in impurity")
-fig.tight_layout()
-from sklearn.inspection import permutation_importance
-
-start_time = time.time()
-result = permutation_importance(
-    forest, X_test, y_test, n_repeats=10, random_state=42, n_jobs=2
-)
-elapsed_time = time.time() - start_time
-print(f"Elapsed time to compute the importances: {elapsed_time:.3f} seconds")
-
-forest_importances_permutation = pd.Series(
-    result.importances_mean, index=df.columns[0:5]
-)
-fig, ax = plt.subplots()
-forest_importances_permutation.plot.bar(yerr=result.importances_std, ax=ax)
-ax.set_title("Feature importances using permutation on full model")
-ax.set_ylabel("Mean accuracy decrease")
-fig.tight_layout()
-plt.show()
-print(
-    forest_importances_permutation.to_frame()
-    .rename(columns={0: "Feature Importance"})
-    .sort_values("Feature Importance", ascending=False)
-)
+    forest_importances_permutation = pd.Series(
+        result.importances_mean, index=df.iloc[:, :-1].columns
+    )
+    fig, ax = plt.subplots()
+    forest_importances_permutation.plot.bar(yerr=result.importances_std, ax=ax)
```

Comment:
Good job plotting the feature importance, it is always good to visualize your results